### ISSUE TITLE
Add `{-# INLINE #-}` pragmas to all instances

### ISSUE THREAD

**Issue by ChShersh** [kowainik/containers-backpack#33]
Some functions are reexported directly from the corresponding libraries. But most of them are implemented in our signatures. To improve performance, it's better to add `{-# INLINE #-}` pragma to every function.

So, for example, if we have:

```haskell
singleton :: Key k => k -> v -> Map k v
singleton = HM.singleton
```

This should be written like this:

```haskell
singleton :: Key k => k -> v -> Map k v
singleton = HM.singleton
{-# INLINE singleton #-}
```

And we should do this for every function in every instance.

**Comment by piq9117** [comment]
Hello! I want to take this on, but what does it mean "all instances"? does it mean all relative folders?

**Comment by ChShersh** [comment]
@piq9117 Hi! Sure, you can start working on this! By all instances I mean _all functions that implement signatures_. It's kinda an analogy between signatures and typeclasses.

**Comment by qnikst** [comment]
It doesn't solve the exact problem but it may worth adding [-fexpose-all-unfoldings](https://downloads.haskell.org/~ghc/latest/docs/html/users_guide/using-optimisation.html?highlight=expose-all#ghc-flag--fexpose-all-unfoldings) flag to the package, that will instruct GHC to add all functions to `.hi` files. However, it has the same effect as adding `{-# INLINABLE #-}` to all functions, so it will not force GHC to inline everything.

**Comment by ChShersh** [comment]
@qnikst So adding `INLINABLE` is better in this case than adding `INLINE`? Or, even easier. just add `-fexpose-all-unfoldings` flag to every package?

**Comment by qnikst** [comment]
Actually I'm not totally sure, ideally, you should benchmark other the applications that use the library in order to see the difference. (Sorry if you already know that). The general situation with GHC solutions that change inlining behaviour is the following. 
GHC just generate a symbol that can be used in the other library, GHC can export unfolding in .hi file, in that case, the function may be specialized at the use site. By default, GHC uses some heuristics for exporting unfoldings of the functions, but usually, it's small for some definition of function size (if I'm not mistaken there is a cost value that GHC measures when compiling, but better to recheck comments in source code).
GHC has following pragmas to modify behaviour:
1. `{-# SPECIALIZE #-}` GHC generates specialization inside your package, but you should know types in advance; then in the other packages when the function will be used with the same types, already specialized version will be used. So no additional work when compiling other functions will be done.
On the other hand, as unfolding may not be there, no further optimizations at the call site are possible.
2. `{-# INLINABLE #-}` tells GHC to export unfolding, so function may be specialized later at the call site, but no work reuse across the different packages is possible. The problem here is that GHC's heuristics may be bad and it will not use the unfolded version, especially in case of the recursive functions, as GHC is really careful there.
3. `{-# INLINE #-}` tells GHC to export unfolding *and* to inline function at the call site. I'm not sure about the behaviour of the GHC when it hits recursive functions. But the main problem here may be too much inlining and overspecialization, as a result, you may get a very large core that works slow. (Sometimes a case with `vector` package).

Up to my mind, the best course of actions will be to add `-fexpose-all-unfoldings` and add `{-#INLINE #-}` to the functions that can definitely be inlined without harm. (It this package it's likely that it's all nonrecursive functions at least.

**Comment by ChShersh** [comment]
@qnikst Thanks for the very useful comments!

**Comment by ChShersh** [comment]
Closed in #36
